File: everything_parsing/21-nullable-first/java/FirstSetsTest.java
import java.util.List;
import java.util.Set;
import java.util.TreeSet;

/** java -cp java/out FirstSetsTest */
public final class FirstSetsTest {

    static FirstSets load(String g) { return new FirstSets(Grammar.parse(g)); }
    static String s(Set<String> set) { return new TreeSet<>(set).toString(); }

    public static void main(String[] args) {
        System.out.println("FirstSetsTest");

        // ── indirect nullability, discovered one layer per round ──
        FirstSets ind = load("""
            %start X
            X -> Y
            Y -> Z
            Z -> z | epsilon
            """);
        Assert.equals(s(ind.nullable), "[X, Y, Z]", "X, Y, Z all nullable (indirectly)");
        // round 0 {}, round 1 {Z}, round 2 {Y,Z}, round 3 {X,Y,Z}, round 4 {X,Y,Z}
        Assert.equals(ind.nullableRounds.size(), 5, "nullable took 3 productive rounds + 1 confirming pass + round 0");
        Assert.equals(s(ind.nullableRounds.get(1)), "[Z]", "round 1 finds only Z");
        Assert.equals(s(ind.nullableRounds.get(2)), "[Y, Z]", "round 2 adds Y");
        Assert.equals(s(ind.firstOf("X")), "[epsilon, z]", "FIRST(X) = { z epsilon }");
        Assert.that(ind.firstIsStable(), "the FIRST result is a fixed point");

        // ── nullable cascade: FIRST(S) reaches past the nullable prefix ──
        FirstSets cas = load("""
            %start S
            S -> A B C d
            A -> a | epsilon
            B -> b | epsilon
            C -> c | epsilon
            """);
        Assert.equals(s(cas.nullable), "[A, B, C]", "A B C nullable; S is not");
        Assert.that(!cas.nullable.contains("S"), "S not nullable (the d)");
        Assert.equals(s(cas.firstOf("S")), "[a, b, c, d]", "FIRST(S) collects a,b,c AND d");
        Assert.that(!cas.firstOf("S").contains("epsilon"), "epsilon NOT in FIRST(S)");

        // ── nullableSeq / firstOfSeq ──
        Assert.that(cas.nullableSeq(List.of("A", "B", "C")), "A B C is a nullable sequence");
        Assert.that(!cas.nullableSeq(List.of("A", "B", "C", "d")), "A B C d is not (d is a terminal)");
        Assert.equals(s(cas.firstOfSeq(List.of("A", "B", "C"))), "[a, b, c, epsilon]",
            "FIRST(A B C) includes epsilon (all nullable)");
        Assert.equals(s(cas.firstOfSeq(List.of("C", "d"))), "[c, d]", "FIRST(C d): c, then d because C nullable");

        // ── expression grammar: circular FIRST needs the iteration ──
        FirstSets e = load("""
            %start E
            E  -> T Ep
            Ep -> + T Ep | epsilon
            T  -> F Tp
            Tp -> * F Tp | epsilon
            F  -> ( E ) | num | id
            """);
        Assert.equals(s(e.nullable), "[Ep, Tp]", "only Ep and Tp are nullable");
        Assert.equals(s(e.firstOf("E")), "[(, id, num]", "FIRST(E) resolved through the E->T->F->(E) cycle");
        Assert.that(e.firstRounds.size() >= 4, "FIRST(E) is empty until round 3 (E<-T<-F chain)");
        Assert.equals(s(e.firstRounds.get(1).get("E")), "[]", "FIRST(E) still empty after round 1");

        // ── cross-check against Module 20's independent implementation ──
        Predict pr = new Predict(e.g);
        for (String nt : e.g.nonterminals)
            Assert.equals(s(e.firstOf(nt)), s(pr.firstOf(nt)), "FIRST(" + nt + ") agrees with Module 20");

        Assert.summary();
    }
}

File: everything_parsing/21-nullable-first/js/firstsets.test.mjs
// node js/firstsets.test.mjs

import { that, equals, summary } from "./assert.mjs";
import { Grammar } from "./grammar.mjs";
import { FirstSets } from "./firstsets.mjs";
import { Predict } from "./predict.mjs";

console.log("firstsets.test.mjs");

const load = g => new FirstSets(Grammar.parse(g));
const s = set => JSON.stringify([...set].sort());

// ── indirect nullability ──
const ind = load(`
%start X
X -> Y
Y -> Z
Z -> z | epsilon
`);
equals(s(ind.nullable), s(new Set(["X", "Y", "Z"])), "X, Y, Z all nullable (indirectly)");
equals(ind.nullableRounds.length, 5, "nullable: round 0 + 3 productive + 1 confirming");
equals(s(ind.nullableRounds[1]), s(new Set(["Z"])), "round 1 finds only Z");
equals(s(ind.nullableRounds[2]), s(new Set(["Y", "Z"])), "round 2 adds Y");
equals(s(ind.firstOf("X")), s(new Set(["epsilon", "z"])), "FIRST(X) = { z epsilon }");
that(ind.firstIsStable(), "the FIRST result is a fixed point");

// ── nullable cascade ──
const cas = load(`
%start S
S -> A B C d
A -> a | epsilon
B -> b | epsilon
C -> c | epsilon
`);
equals(s(cas.nullable), s(new Set(["A", "B", "C"])), "A B C nullable; S is not");
that(!cas.nullable.has("S"), "S not nullable (the d)");
equals(s(cas.firstOf("S")), s(new Set(["a", "b", "c", "d"])), "FIRST(S) collects a,b,c AND d");
that(!cas.firstOf("S").has("epsilon"), "epsilon NOT in FIRST(S)");

that(cas.nullableSeq(["A", "B", "C"]), "A B C is a nullable sequence");
that(!cas.nullableSeq(["A", "B", "C", "d"]), "A B C d is not");
equals(s(cas.firstOfSeq(["A", "B", "C"])), s(new Set(["a", "b", "c", "epsilon"])),
  "FIRST(A B C) includes epsilon");
equals(s(cas.firstOfSeq(["C", "d"])), s(new Set(["c", "d"])), "FIRST(C d): c, then d because C nullable");

// ── expression grammar ──
const e = load(`
%start E
E  -> T Ep
Ep -> + T Ep | epsilon
T  -> F Tp
Tp -> * F Tp | epsilon
F  -> ( E ) | num | id
`);
equals(s(e.nullable), s(new Set(["Ep", "Tp"])), "only Ep and Tp are nullable");
equals(s(e.firstOf("E")), s(new Set(["(", "id", "num"])), "FIRST(E) resolved through the cycle");
that(e.firstRounds.length >= 4, "FIRST(E) empty until round 3");
equals(s(e.firstRounds[1].get("E")), s(new Set()), "FIRST(E) still empty after round 1");

// ── cross-check against Module 20 ──
const pr = new Predict(e.g);
for (const nt of e.g.nonterminals)
  equals(s(e.firstOf(nt)), s(pr.firstOf(nt)), `FIRST(${nt}) agrees with Module 20`);

summary();
